lib/libzfs/libzfs_util.c

```diff
@@ -2178,3 +2181,79 @@ zprop_iter(zprop_func func, void *cb, boolean_t show_all, boolean_t ordered,
 {
 	return (zprop_iter_common(func, cb, show_all, ordered, type));
 }
+
+/*
+ * zfs_get_hole_count() retrieves the number of holes (blocks which are
+ * zero-filled) in the specified file using the FS_IOC_FIEMAP ioctl.  It
+ * also optionally fetches the block size when bs is non-NULL.  With hole
+ * count and block size the full space consumed by the holes of a file can
+ * be calculated.
+ *
+ * On success, zero is returned, the count argument is set to the number of
+ * unallocated blocks (holes), and the bs argument is set to the block size
+ * (if it is not NULL). On error, a non-zero errno is returned and the values
+ * in count and bs are undefined.
+ */
+int
+zfs_get_hole_count(const char *path, uint64_t *count, uint64_t *bs)
+{
+	struct fiemap *fiemap;
+	struct stat64 ss;
+	uint64_t fill;
+	int fd, error;
+
+	fd = open(path, O_RDONLY);
+	if (fd == -1)
+		return (errno);
+
+	fiemap = calloc(1, sizeof (struct fiemap));
+	if (fiemap == NULL) {
+		error = errno;
+		(void) close(fd);
+		return (error);
+	}
+
+	fiemap->fm_start = 0;
+	fiemap->fm_length = FIEMAP_MAX_OFFSET;
+	fiemap->fm_flags = FIEMAP_FLAG_NOMERGE;
+	fiemap->fm_extent_count = 0;
+	fiemap->fm_mapped_extents = 0;
+
+	error = ioctl(fd, FS_IOC_FIEMAP, fiemap);
+	if (error < 0) {
+		error = errno;
+		free(fiemap);
+		(void) close(fd);
+		return (error);
+	}
+
+	fill = fiemap->fm_mapped_extents;
+	free(fiemap);
+
+	if (fstat64(fd, &ss) == -1) {
+		error = errno;
+		(void) close(fd);
+		return (error);
+	}
+
+	if (ss.st_blksize == 0) {
+		(void) close(fd);
+		return (EINVAL);
+	}
+
+	/*
+	 * The number of blocks times the block size may exceed the file size
+	 * when there are pending dirty blocks which have not be written.
```

Comment:
I believe it can for this special case where `FIEMAP_FLAG_NOMERGE` is passed and  we return the fill count plus the pending dirty blocks.  The dirty blocks might overwrite an existing block so we over report slightly.  This might also be possible when the `copies` property is set larger than 1.

Now that's fine since if the actual extents are requested you'll just allocate a slightly larger buffer than needed.  The returned extents will be accurate.